File: app/src/main/java/com/walagran/wwf/ui/common/KeyboardFragment.java
package com.walagran.wwf.ui.common;

import android.os.Bundle;

import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.walagran.wwf.R;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link KeyboardFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class KeyboardFragment extends Fragment {

    KeyboardEventListener keyboardEventListener;

    public KeyboardFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @param keyboardEventListener Listens to the keyboard events and acts accordingly.
     * @return A new instance of fragment KeyboardFragment.
     */
    public static KeyboardFragment newInstance(KeyboardEventListener keyboardEventListener) {
        KeyboardFragment fragment = new KeyboardFragment();
        fragment.setKeyBoardEventListener(keyboardEventListener);
        return fragment;
    }

    public void setKeyBoardEventListener(KeyboardEventListener keyboardEventListener) {
        this.keyboardEventListener = keyboardEventListener;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView = inflater.inflate(R.layout.fragment_keyboard, container, false);

        for (char alpha = 'a'; alpha <= 'z'; alpha++) {
            char passedAlphabet = alpha;
            rootView
                    .findViewById(rootView
                            .getResources()
                            .getIdentifier("key_"+alpha, "id", rootView
                                    .getContext()
                                    .getPackageName()))
                    .setOnClickListener(view -> keyboardEventListener
                            .onAlphaKeyPressed(passedAlphabet));
        }

        return rootView;
    }
}

File: app/src/main/java/com/walagran/wwf/ui/common/KeyboardFragment.kt
package com.walagran.wwf.ui.common

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import androidx.fragment.app.Fragment
import com.walagran.wwf.R

/**
 * A dumb QWERTY keyboard with enter and back key.
 */
class KeyboardFragment : Fragment() {
    private lateinit var keyboardEventListener: KeyboardEventListener

    fun setKeyBoardEventListener(keyboardEventListener: KeyboardEventListener) {
        this.keyboardEventListener = keyboardEventListener
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        // Inflate the layout for this fragment
        val rootView = inflater.inflate(
            R.layout.fragment_keyboard,
            container,
            false)
        var alpha = 'a'
        while (alpha <= 'z') {
            val cachedAlphabet = alpha
            rootView
                .findViewById<Button>(rootView
                    .resources
                    .getIdentifier("key_$cachedAlphabet", "id", rootView
                        .context
                        .packageName))
                .setOnClickListener {
                    keyboardEventListener.onAlphaKeyPressed(cachedAlphabet.uppercaseChar())
                }
            alpha++
        }
        rootView
            .findViewById<Button>(R.id.key_back)
            .setOnClickListener { keyboardEventListener.onBackKeyPressed() }
        rootView
            .findViewById<Button>(R.id.key_enter)
            .setOnClickListener {
                val characterStateMap =
                    keyboardEventListener.onEnterKeyPressed()
                characterStateMap.forEach { (character, letterState) ->
                    rootView.findViewById<Button>(rootView.resources.getIdentifier(
                        "key_$character",
                        "id",
                        rootView.context.packageName))
                        .setBackgroundColor(rootView.resources!!.getColor(
                            letterState.color(),
                            rootView.context.theme))
                }
            }
        return rootView
    }

    companion object {
        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @param keyboardEventListener Listens to the keyboard events and acts
         * accordingly.
         * @return A new instance of fragment KeyboardFragment.
         */
        @JvmStatic
        fun newInstance(keyboardEventListener: KeyboardEventListener): KeyboardFragment {
            val fragment = KeyboardFragment()
            fragment.setKeyBoardEventListener(keyboardEventListener)
            return fragment
        }
    }
}